### PR TITLE
Add string interpolation support

### PR DESCRIPTION
**Pull request by Shan1024**
This PR adds string interpolation support to Ballerina.

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/parser/antlr4/BLangAntlr4Listener.java [hunk 1]
```diff
@@ -2678,6 +2691,58 @@ public void exitXmlQualifiedName(XmlQualifiedNameContext ctx) {
...
+    }
+
+    @Override
+    public void exitStringTemplateContent(StringTemplateContentContext ctx) {
+        if (ctx.exception != null) {
+            return;
+        }
+        List<TerminalNode> nodes = ctx.StringTemplateExpressionStart();
+        StringTemplateTextContext endTextNode = ctx.stringTemplateText();
+        String endingText = endTextNode == null ? null : StringEscapeUtils.unescapeJava(endTextNode.getText());
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/parser/antlr4/BLangAntlr4Listener.java]: Here unescape if the 'verboseMode' flag is not set. 

'verboseMode' flag is set by the composer to persist white spaces in the code.

Refer the 'createBasicLiteral' method in this class.

File: docs/grammar/BallerinaLexer.g4 [hunk 2]
```diff
@@ -675,3 +679,41 @@ XMLCommentSpecialSequence
...
+    ;
+
+StringTemplateExpressionStart
+    :   StringTemplateText? ExpressionStart            -> pushMode(DEFAULT_MODE)
+    ;
+
+StringTemplateText
+    :   StringTemplateBracesSequence? (StringTemplateStringChar StringTemplateBracesSequence?)+
+    |   StringTemplateBracesSequence (StringTemplateStringChar StringTemplateBracesSequence?)*
+    ;
```
Comment (sameera-jayasoma) [inline docs/grammar/BallerinaLexer.g4]: Looks like the second product is not needed. How about the following grammar rule?

StringTemplateText
 +    :   StringTemplateBracesSequence? (StringTemplateStringChar StringTemplateBracesSequence?)*

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/parser/antlr4/BLangAntlr4Listener.java [hunk 3]
```diff
@@ -2678,6 +2691,58 @@ public void exitXmlQualifiedName(XmlQualifiedNameContext ctx) {
...
+
+        if (nodes.size() == 0) {
+            modelBuilder.createStringTemplateLiteral(getCurrentLocation(ctx), null, endingText);
+            return;
+        }
+
+        String[] templateStrLiterals = new String[nodes.size()];
+        int i = 0;
+        for (TerminalNode node : nodes) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/parser/antlr4/BLangAntlr4Listener.java]: Can these nodes be null?

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/StringTemplateLiteral.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import org.ballerinalang.model.NodeVisitor;
+import org.ballerinalang.model.WhiteSpaceDescriptor;
+import org.ballerinalang.model.types.BTypes;
+
+/**
+ * {@link StringTemplateLiteral} represents a string template literal in Ballerina.
+ *
+ * @since 0.92
+ */
+public class StringTemplateLiteral extends AbstractExpression {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/StringTemplateLiteral.java]: You can extend from the NaryExpression class.

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 5]
```diff
@@ -2456,6 +2457,48 @@ public void visit(LambdaExpression lambdaExpr) {
     }
 
     @Override
+    public void visit(StringTemplateLiteral stringTemplateLiteral) {
+        Expression[] items = stringTemplateLiteral.getItems();
+        List<Expression> newItems = new ArrayList<>();
+        // Consecutive non-string type items are converted to string, and combined together using binary add
+        // expressions.
+        Expression addExpr = null;
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: Create the complete BinaryAddExpression first and then evaluate the complete expression.

File: docs/grammar/BallerinaLexer.g4 (hunk 2 above)
Comment (Shan1024) [inline docs/grammar/BallerinaLexer.g4]: We have two rules here because the suggested rule can match an empty string.

**Pull request merged**